plugins/keysend.c

```diff
@@ -2,17 +2,186 @@
 #include <ccan/array_size/array_size.h>
 #include <ccan/tal/str/str.h>
 #include <plugins/libplugin.h>
+#include <plugins/libplugin-pay.h>
 #include <wire/gen_onion_wire.h>
 
 #define PREIMAGE_TLV_TYPE 5482373484
 #define KEYSEND_FEATUREBIT 55
+static unsigned int maxdelay_default;
+static struct node_id my_id;
+
+/*****************************************************************************
+ * Keysend modifier
+ * ================
+ *
+ * The keysend modifier adds the payment preimage to the TLV payload. This
+ * enables the recipient to accept the payment despite it not correspondin to
+ * an invoice that the recipient created. Keysend does not provide any proof
+ * or payment, but does not require an out-of-band communication round to get
+ * an invoice first.
+ */
+
+/* FIXME: If we have more than one plugin using keysend we can move this to
+ * libplugin-pay.c */
+
+struct keysend_data {
+	struct preimage preimage;
+};
+
+REGISTER_PAYMENT_MODIFIER_HEADER(keysend, struct keysend_data);
+
+static struct keysend_data *keysend_init(struct payment *p)
+{
+	struct keysend_data *d;
+	struct sha256 payment_hash;
+	if (p->parent == NULL) {
+		/* If we are the root payment we generate a random preimage
+		 * and populate the preimage field in the keysend_data and the
+		 * payment_hash in the payment. */
+		d = tal(p, struct keysend_data);
+		randombytes_buf(&d->preimage, sizeof(d->preimage));
+		ccan_sha256(&payment_hash, &d->preimage, sizeof(d->preimage));
+		p->payment_hash = tal_dup(p, struct sha256, &payment_hash);
+		return d;
+	} else {
+		/* If we are a child payment (retry or split) we copy the
+		 * parent's information, since the payment_hash needs to match
+		 * in order to be collated at the recipient. */
+		return payment_mod_keysend_get_data(p->parent);
+	}
+}
+
+static void keysend_cb(struct keysend_data *d, struct payment *p) {
+	struct route_hop *last_hop;
+	struct createonion_hop *last_payload;
+	size_t hopcount;
+	u8 *raw_preimage;
+
+	if (p->step == PAYMENT_STEP_GOT_ROUTE) {
+		last_hop = &p->route[tal_count(p->route) - 1];
+		if (last_hop->style == ROUTE_HOP_LEGACY)
+			payment_fail(p,
+				     "Destination does not support receiving "
+				     "keysend payments (not a TLV hop)");
+	}
+
+	if (p->step != PAYMENT_STEP_ONION_PAYLOAD)
+		return payment_continue(p);
+
+	raw_preimage = tal_dup_arr(p->createonion_request, u8, d->preimage.r,
+				   sizeof(d->preimage), 0);
+
+	hopcount = tal_count(p->createonion_request->hops);
+	last_payload = &p->createonion_request->hops[hopcount - 1];
+	tlvstream_set_raw(&last_payload->tlv_payload->fields, PREIMAGE_TLV_TYPE,
+			  take(raw_preimage));
```

Comment:
Aside: maybe tlvstream_set_raw should take const void *, size_t.  That would avoid this temp allocation.